libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,579 @@
+	// then we want the variable tmp in the slot at offset 2 in the layout before the operation.
+	vector<optional<StackSlot>> idealLayout(_post.size(), nullopt);
+	for (auto const& [slot, idealPosition]: ranges::zip_view(_post, layout))
+		if (PreviousSlot* previousSlot = std::get_if<PreviousSlot>(&idealPosition))
+			idealLayout.at(previousSlot->slot) = slot;
+
+	// The tail of layout must have contained the operation outputs and will not have been assigned slots in the last loop.
+	while (!idealLayout.empty() && !idealLayout.back())
+		idealLayout.pop_back();
+
+	yulAssert(idealLayout.size() == previousLayoutSize, "");
+
+	return idealLayout | ranges::views::transform([](optional<StackSlot> s) {
+		yulAssert(s, "");
+		return *s;
+	}) | ranges::to<Stack>;
+}
+}
+
+Stack StackLayoutGenerator::propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation)
+{
+	// This is a huge tradeoff between code size, gas cost and stack size.
+	auto generateSlotOnTheFly = [&](StackSlot const&) {
+		//return stack.size() > 12 && canBeFreelyGenerated(_slot);
+		// return canBeFreelyGenerated(_slot);
+		return false;
+	};
+
+	// Determine the ideal permutation of the slots in _exitLayout that are not operation outputs (and not to be
+	// generated on the fly), s.t. shuffling the stack+_operation.output to _exitLayout is cheap.
+	Stack stack = createIdealLayout(_operation.output, _exitStack, generateSlotOnTheFly);
+
+	// Make sure the resulting previous slots do not overlap with any assignmed variables.
+	if (auto const* assignment = get_if<CFG::Assignment>(&_operation.operation))
+		for (auto& stackSlot: stack)
+			if (auto const* varSlot = get_if<VariableSlot>(&stackSlot))
+				yulAssert(!util::findOffset(assignment->variables, *varSlot), "");
+
+	// Since stack+_operation.output can be easily shuffled to _exitLayout, the desired layout before the operation
+	// is stack+_operation.input;
+	stack += _operation.input;
+
+	// Store the exact desired operation entry layout. The stored layout will be recreated by the code transform
+	// before executing the operation. However, this recreation can produce slots that can be freely generated or
+	// are duplicated, i.e. we can compress the stack afterwards without causing problems for code generation later.
+	m_layout.operationEntryLayout[&_operation] = stack;
+
+	// Remove anything from the stack top that can be freely generated or dupped from deeper on the stack.
+	while (!stack.empty())
+	{
+		if (canBeFreelyGenerated(stack.back()))
+			stack.pop_back();
+		else if (auto offset = util::findOffset(stack | ranges::views::reverse | ranges::views::drop(1), stack.back()))
+		{
+			if (*offset + 2 < 16)
+				stack.pop_back();
+			else
+				break;
+		}
+		else
+			break;
+	}
+
+	// TODO: there may be a better criterion than overall stack size.
+	if (stack.size() > 12)
+		// Deduplicate and remove slots that can be freely generated.
+		stack = compressStack(move(stack));
```

Comment:
This might need more tweaking - it doesn't hurt for our test cases, but it makes it rather easy for the fuzzer to create stack-too-deep errors. Just removing it actually also passes all tests, but it will probably rely on much more stack-to-memory that way... the sanest choice is probably to do something like this, but only if it fixes a stack-too-deep that would otherwise occur...